cpp/tests/io/algo/scan_select_if.cuh

```diff
@@ -0,0 +1,432 @@
+    index_tile_state  //
+  };
+
+  agent_instance.consume_range(do_scatter, num_tiles, start_tile);
+}
+
+// ===== Policy ====================================================================================
+
+template <typename InputIterator_,
+          typename OutputCountIterator_,
+          typename OutputIterator_,
+          typename ScanOperator_,
+          typename PredicateOperator_>
+struct policy {
+  enum : uint32_t {
+    THREADS_PER_INIT_BLOCK = 128,
+    THREADS_PER_BLOCK      = 128,
+    ITEMS_PER_THREAD       = 16,
+    ITEMS_PER_TILE         = ITEMS_PER_THREAD * THREADS_PER_BLOCK,
+  };
+
+  using InputIterator       = InputIterator_;
+  using OutputIterator      = OutputIterator_;
+  using OutputCountIterator = OutputCountIterator_;
+
+  using ScanOperator = ScanOperator_;
+  using PredOperator = PredicateOperator_;
+
+  using Offset      = typename std::iterator_traits<InputIterator>::difference_type;
+  using Input       = typename std::iterator_traits<InputIterator>::value_type;
+  using OutputCount = typename std::iterator_traits<OutputCountIterator>::value_type;
+  using OutputValue = typename std::iterator_traits<OutputIterator>::value_type;
+
+  // Items Load
+
+  using ItemsBlockLoad = cub::BlockLoad<  //
+    Input,
+    THREADS_PER_BLOCK,
+    ITEMS_PER_THREAD,
+    cub::BlockLoadAlgorithm::BLOCK_LOAD_DIRECT>;
+
+  // Items Scan
+
+  using ItemsTileState      = cub::ScanTileState<Input>;
+  using ItemsPrefixCallback = cub::TilePrefixCallbackOp<Input, ScanOperator, ItemsTileState>;
+
+  using ItemsBlockScan = cub::BlockScan<  //
+    Input,
+    THREADS_PER_BLOCK,
+    cub::BlockScanAlgorithm::BLOCK_SCAN_WARP_SCANS>;
+
+  // Index Scan
+
+  using IndexTileState      = cub::ScanTileState<uint32_t>;
+  using IndexPrefixCallback = cub::TilePrefixCallbackOp<uint32_t, cub::Sum, IndexTileState>;
+  using IndexBlockScan      = cub::BlockScan<  //
+    uint32_t,
+    THREADS_PER_BLOCK,
+    cub::BlockScanAlgorithm::BLOCK_SCAN_WARP_SCANS>;
+};
+
+// ===== Entry =====================================================================================
+
+template <typename InputIterator,
+          typename OutputCountIterator,
+          typename OutputIterator,
+          typename ScanOperator,
+          typename PredOperator>
+void scan_select_if(  //
+  void* d_temp_storage,
+  size_t& temp_storage_bytes,
+  InputIterator d_in,
+  OutputCountIterator d_count_out,
+  OutputIterator d_out,
+  uint32_t num_items,
+  ScanOperator scan_op,
+  PredOperator pred_op,
+  bool do_initialize,
+  bool do_scatter,
+  cudaStream_t stream = 0)
+{
+  CUDF_FUNC_RANGE();
+
+  using Policy =
+    policy<InputIterator, OutputCountIterator, OutputIterator, ScanOperator, PredOperator>;
+
+  uint32_t num_tiles = ceil_div(num_items, Policy::ITEMS_PER_TILE);
+
+  // calculate temp storage requirements
+
+  void* allocations[2];
+  size_t allocation_sizes[2];
+
+  CUDA_TRY(Policy::ItemsTileState::AllocationSize(num_tiles, allocation_sizes[0]));
```

Comment:
Why is a non-CUDART function wrapped in `CUDA_TRY`?

Reply:
`ItemsTileState` is an alias for `cub::ScanTileState`, who's static `AllocationSize` function returns `cudaError_t`. The implementation might as well be `constexpr`, but the signature of the method suggests we should check for errors... :shrug:

Am I missing something? I was under the impression we should use `CUDA_TRY` for any functions returning `cudaError_t`.

Reply:
I didn't know that CUB returns `cudaError_t`. Seems weird -- what if there are errors in CUB that are not CUDA errors.